// generator-ng-component/generators/app/index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _yeomanGenerator = require('yeoman-generator');

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Generator = function (_Base) {
  _inherits(Generator, _Base);

  function Generator() {
    _classCallCheck(this, Generator);

    return _possibleConstructorReturn(this, (Generator.__proto__ || Object.getPrototypeOf(Generator)).apply(this, arguments));
  }

  _createClass(Generator, [{
    key: 'initializing',
    value: function initializing() {
      if (!this.options['skip-message']) {
        this.log(_chalk2.default.magenta('You\'re using the fantastic NgComponent generator.\n'));
        this.log(_chalk2.default.magenta('Initializing yo-rc.json configuration.\n'));
      }
    }
  }, {
    key: 'configuring',
    value: function configuring() {
      var config = {
        routeDirectory: this.options.routeDirectory || 'app/components/',
        directiveDirectory: this.options.directiveDirectory || 'app/components/',
        componentDirectory: this.options.componentDirectory || 'app/components/',
        filterDirectory: this.options.filterDirectory || 'app/components/',
        serviceDirectory: this.options.serviceDirectory || 'app/components/',
        basePath: this.options.basePath || 'app',
        moduleName: this.options.moduleName || '',
        modulePrompt: {}.hasOwnProperty.call(this.options, 'modulePrompt') ? this.options.modulePrompt : true,
        filters: this.options.filters || ['uirouter', 'jasmine'],
        extensions: this.options.extensions || ['js', 'html', 'scss'],
        directiveSimpleTemplates: this.options.directiveSimple || '',
        directiveComplexTemplates: this.options.directiveComplex || '',
        filterTemplates: this.options.filter || '',
        serviceTemplates: this.options.service || '',
        factoryTemplates: this.options.factory || '',
        controllerTemplates: this.options.controller || '',
        componentTemplates: this.options.component || '',
        decoratorTemplates: this.options.decorator || '',
        providerTemplates: this.options.provider || '',
        routeTemplates: this.options.route || ''
      };

      if (this.options.forceConfig) {
        this.config.set(config);
        this.config.forceSave();
      } else {
        this.config.defaults(config);
      }
    }
  }]);

  return Generator;
}(_yeomanGenerator.Base);

module.exports = Generator;
