// generator-ng-component/generators/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rewriteFile = rewriteFile;
exports.rewrite = rewrite;
exports.appName = appName;
exports.copyTemplates = copyTemplates;
exports.relativeUrl = relativeUrl;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = {
  rewrite: rewrite,
  rewriteFile: rewriteFile,
  appName: appName,
  copyTemplates: copyTemplates,
  relativeUrl: relativeUrl
};

function rewriteFile(args) {
  args.path = args.path || process.cwd();
  var fullPath = _path2.default.join(args.path, args.file);

  args.haystack = _fs2.default.readFileSync(fullPath, 'utf8');
  var body = rewrite(args);

  _fs2.default.writeFileSync(fullPath, body);
}

function escapeRegExp(str) {
  return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
}

function rewrite(args) {
  // check if splicable is already in the body text
  var re = new RegExp(args.splicable.map(function (line) {
    return 's*' + escapeRegExp(line);
  }).join('\n'));

  if (re.test(args.haystack)) {
    return args.haystack;
  }

  var lines = args.haystack.split('\n');

  var otherwiseLineIndex = 0;
  lines.forEach(function (line, i) {
    if (line.indexOf(args.needle) !== -1) {
      otherwiseLineIndex = i;
    }
  });

  var spaces = 0;
  while (lines[otherwiseLineIndex].charAt(spaces) === ' ') {
    spaces += 1;
  }

  var spaceStr = '';
  while ((spaces -= 1) >= 0) {
    spaceStr += ' ';
  }

  lines.splice(otherwiseLineIndex, 0, args.splicable.map(function (line) {
    return spaceStr + line;
  }).join('\n'));

  return lines.join('\n');
}

function appName(self) {
  var counter = 0;
  var suffix = self.options['app-suffix'];
  // Have to check this because of generator bug #386
  process.argv.forEach(function (val) {
    if (val.indexOf('--app-suffix') > -1) {
      counter++;
    }
  });
  if (counter === 0 || typeof suffix === 'boolean' && suffix) {
    suffix = 'App';
  }
  return suffix ? self.lodash.classify(suffix) : '';
}

function createFileName(template, name) {
  // Find matches for parans
  var filterMatches = template.match(/\(([^)]+)\)/g);
  var filter = '';
  if (filterMatches) {
    filter = filterMatches[0].replace('(', '').replace(')', '');
    template = template.replace(filterMatches[0], '');
  }

  return { name: template.replace('name', name), filter: filter };
}

function templateIsUsable(processedName, self) {
  var filters = self.config.get('filters') || [];
  var include = true;

  if (processedName.filter && filters.indexOf(processedName.filter) === -1) {
    include = false;
  }

  var index = processedName.name.lastIndexOf('.');
  var ext = processedName.name.slice(index + 1);
  var extensions = self.config.get('extensions') || [];

  return extensions.indexOf(ext) >= 0 && include;
}

function copyTemplates(self, type, templateDir, configName) {
  templateDir = templateDir || _path2.default.join(self.sourceRoot(), type);
  configName = configName || type + 'Templates';

  if (self.config.get(configName)) {
    templateDir = _path2.default.join(process.cwd(), self.config.get(configName));
  }

  var useTypeScript = self.config.get('extensions').indexOf('ts') > -1;

  var files = (0, _lodash2.default)(_fs2.default.readdirSync(templateDir)).map(function (template) {
    var processedName = createFileName(template, self.name);
    var fileName = processedName.name;
    var templateFile = _path2.default.join(templateDir, template);
    var dest = _path2.default.join(self.dir, fileName);
    if (useTypeScript && dest.indexOf('.json') === -1) {
      dest = dest.replace('.js', '.ts');
    }

    return {
      processedName: processedName,
      fileName: fileName,
      templateFile: templateFile,
      dest: dest
    };
  }).filter(function (templateObj) {
    return templateIsUsable(templateObj.processedName, self);
  }).value();

  files.forEach(function (templateObj) {
    self.fs.copyTpl(templateObj.templateFile, templateObj.dest, self);
  });
}

function relativeUrl(basePath, targetPath) {
  var relativePath = _path2.default.relative(basePath, targetPath);
  return relativePath.split(_path2.default.sep).join('/');
}
